### PR TITLE
Reduce federation replication traffic

### PR DESCRIPTION
**Pull request by erikjohnston**
This is mainly done by moving the calculation of where to send presence updates from the presence handler to the transaction queue, so we only need to send the presence event (and not the destinations) across the replication connection. Before we were duplicating by sending the full state across once per destination.

### CODE REVIEW

File: synapse/federation/transaction_queue.py [hunk 1]
```diff
@@ -224,17 +227,95 @@ def _send_pdu(self, pdu, destinations):
...
+            hosts = yield self.store.get_hosts_in_room(room_id)
+            hosts_and_states.append((hosts, local_states))
+
+        for user_id, states in users_to_states.items():
+            local_states = filter(lambda s: self.is_mine_id(s.user_id), states)
+            if not local_states:
+                continue
+
+            host = get_domain_from_id(user_id)
+            hosts_and_states.append(([host], local_states))
```
Comment (erikjohnston) [inline synapse/federation/transaction_queue.py]: The above is copied verbatim from `_get_interested_parties`

File: synapse/federation/transaction_queue.py [hunk 2]
```diff
@@ -224,17 +227,95 @@ def _send_pdu(self, pdu, destinations):
...
+                if not self.can_send_to(destination):
+                    continue
+
+                self.pending_presence_by_dest.setdefault(
+                    destination, {}
+                ).update({
+                    state.user_id: state for state in states
+                })
+
+                preserve_fn(self._attempt_new_transaction)(destination)
```
Comment (erikjohnston) [inline synapse/federation/transaction_queue.py]: The above is copied verbatim from the old `send_presence` function

File: synapse/federation/transaction_queue.py [hunk 3]
```diff
@@ -224,17 +227,95 @@ def _send_pdu(self, pdu, destinations):
                 self._attempt_new_transaction, destination
             )
 
-    def send_presence(self, destination, states):
-        if not self.can_send_to(destination):
+    @preserve_fn
```
Comment (erikjohnston) [inline synapse/federation/transaction_queue.py]: Is this an acceptable thing to do style wise?

File: synapse/federation/transaction_queue.py [hunk 4]
```diff
@@ -78,6 +78,7 @@ def __init__(self, hs):
         self.pending_edus_by_dest = edus = {}
 
         # Presence needs to be separate as we send single aggragate EDUs
+        self.pending_presence = {}
```
Comment (richvdh) [inline synapse/federation/transaction_queue.py]: could you document what all three of these fields mean and their types?

File: synapse/federation/transaction_queue.py [hunk 5]
```diff
@@ -224,17 +227,95 @@ def _send_pdu(self, pdu, destinations):
                 self._attempt_new_transaction, destination
             )
 
-    def send_presence(self, destination, states):
-        if not self.can_send_to(destination):
+    @preserve_fn
+    @defer.inlineCallbacks
+    def send_presence(self, states):
+        """Send the new presence states to the appropriate destinations.
```
Comment (richvdh) [inline synapse/federation/transaction_queue.py]: "Start sending" ... "if we are not already sending updates" or something

File: synapse/federation/transaction_queue.py [hunk 6]
```diff
@@ -224,17 +227,95 @@ def _send_pdu(self, pdu, destinations):
...
+
+        Args:
+            states (list(UserPresenceState))
+        """
+        # First we look up the rooms each user is in (as well as any explicit
+        # subscriptions), then for each distinct room we look up the remote
+        # hosts in those rooms.
+        room_ids_to_states = {}
+        users_to_states = {}
+        for state in states.itervalues():
```
Comment (richvdh) [inline synapse/federation/transaction_queue.py]: please let's not c&p `get_interested_parties` here. How about moving this lump of code to the store or something?

(For super-mega-bonus points, make it a separate PR which precedes this one...)

File: synapse/federation/transaction_queue.py [hunk 7]
```diff
@@ -224,17 +227,95 @@ def _send_pdu(self, pdu, destinations):
...
+            for room_id in room_ids:
+                room_ids_to_states.setdefault(room_id, []).append(state)
+
+            plist = yield self.store.get_presence_list_observers_accepted(
+                state.user_id,
+            )
+            for u in plist:
+                users_to_states.setdefault(u, []).append(state)
+
+        hosts_and_states = []
```
Comment (richvdh) [inline synapse/federation/transaction_queue.py]: comment on what this is going to be, please.

File: synapse/federation/transaction_queue.py [hunk 8]
```diff
@@ -224,17 +227,95 @@ def _send_pdu(self, pdu, destinations):
...
+        for room_id, states in room_ids_to_states.items():
+            local_states = filter(lambda s: self.is_mine_id(s.user_id), states)
```
Comment (richvdh) [inline synapse/federation/transaction_queue.py]: again, why are we filtering, and why would updates for remote users be getting in here? (and could we filter *before* deriving `room_ids_to_states` and `users_to_states` to avoid some work? happy if you want to leave this for now to avoid changing too much at once)

**Comment by erikjohnston** [comment]
I've moved the `get_interested_*` functions out into standalone functions in the `handlers/presence.py` file, as it doesn't feel like the sort of thing that belongs in the storage layer

**Comment by richvdh** [comment]
> I've moved the get_interested_* functions out into standalone functions in the handlers/presence.py file, as it doesn't feel like the sort of thing that belongs in the storage layer

Fine by me. Looks much better now.

File: synapse/handlers/presence.py [hunk 9]
```diff
@@ -1318,3 +1290,77 @@ def handle_update(prev_state, new_state, is_mine, wheel_timer, now):
...
+        destination
+    """
+    hosts_and_states = []  # Final result to return
+
+    # First we look up the rooms each user is in (as well as any explicit
+    # subscriptions), then for each distinct room we look up the remote
+    # hosts in those rooms.
+    room_ids_to_states = {}
+    users_to_states = {}
+    for state in states.itervalues():
```
Comment (erikjohnston) [inline synapse/handlers/presence.py]: Heh, for some reason I got it stuck in my head that they were different
